File: internal/adapter/oidc/oidc.go
package oidcadapter

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/coreos/go-oidc/v3/oidc"
	"golang.org/x/oauth2"

	"github.com/yamahigashi/lore-auth-bridge/internal/core/model"
	"github.com/yamahigashi/lore-auth-bridge/internal/core/ports"
)

type Provider struct {
	id                     string
	profile                string
	displayName            string
	issuer                 string
	oauth2Config           oauth2.Config
	verifier               *oidc.IDTokenVerifier
	claimMapping           map[string]string
	subjectStrategy        string
	requiredTenantID       string
	emailBinding           string
	pkce                   string
	allowedEmailDomainList []string
	allowedHostedDomains   map[string]struct{}
	personalAccounts       string
}

type Config struct {
	ProviderID           string
	Profile              string
	DisplayName          string
	Issuer               string
	ClientID             string
	ClientSecret         string
	RedirectURL          string
	Scopes               []string
	ClaimMapping         map[string]string
	SubjectStrategy      string
	RequiredTenantID     string
	EmailBinding         string
	PKCE                 string
	AllowedEmailDomains  []string
	AllowedHostedDomains []string
	PersonalAccounts     string
}

func New(ctx context.Context, cfg Config) (*Provider, error) {
	if cfg.ProviderID == "" {
		return nil, fmt.Errorf("oidc: provider id is required")
	}
	if cfg.Issuer == "" {
		return nil, fmt.Errorf("oidc: issuer is required")
	}
	if cfg.ClientID == "" {
		return nil, fmt.Errorf("oidc: client id is required")
	}
	profile := strings.TrimSpace(cfg.Profile)
	personalAccounts := strings.TrimSpace(cfg.PersonalAccounts)
	switch personalAccounts {
	case "", "allow", "deny":
	default:
		return nil, fmt.Errorf("oidc: personal_accounts %q is unknown", cfg.PersonalAccounts)
	}
	if personalAccounts != "" && profile != "google" {
		return nil, fmt.Errorf("oidc: personal_accounts is only valid for google profile")
	}
	provider, err := oidc.NewProvider(ctx, cfg.Issuer)
	if err != nil {
		return nil, fmt.Errorf("oidc: discover provider %q: %w", cfg.Issuer, err)
	}
	var discoveryClaims struct {
		Issuer string `json:"issuer"`
	}
	if err := provider.Claims(&discoveryClaims); err != nil {
		return nil, fmt.Errorf("oidc: decode discovery claims for %q: %w", cfg.Issuer, err)
	}
	issuer := strings.TrimSpace(discoveryClaims.Issuer)
	if issuer == "" {
		issuer = cfg.Issuer
	}
	scopes := append([]string(nil), cfg.Scopes...)
	if len(scopes) == 0 {
		scopes = []string{oidc.ScopeOpenID, "email", "profile"}
	}
	allowedEmailDomainList := normalizeDomainList(cfg.AllowedEmailDomains)
	return &Provider{
		id:          cfg.ProviderID,
		profile:     profile,
		displayName: displayNameOrDefault(cfg.DisplayName, cfg.ProviderID),
		issuer:      issuer,
		oauth2Config: oauth2.Config{
			ClientID:     cfg.ClientID,
			ClientSecret: cfg.ClientSecret,
			RedirectURL:  cfg.RedirectURL,
			Scopes:       scopes,
			Endpoint:     provider.Endpoint(),
		},
		verifier:               provider.Verifier(&oidc.Config{ClientID: cfg.ClientID}),
		claimMapping:           copyClaimMapping(cfg.ClaimMapping),
		subjectStrategy:        defaultString(cfg.SubjectStrategy, "oidc_sub"),
		requiredTenantID:       strings.TrimSpace(cfg.RequiredTenantID),
		emailBinding:           defaultString(cfg.EmailBinding, "disabled"),
		pkce:                   strings.TrimSpace(cfg.PKCE),
		allowedEmailDomainList: allowedEmailDomainList,
		allowedHostedDomains:   normalizeDomains(cfg.AllowedHostedDomains),
		personalAccounts:       personalAccounts,
	}, nil
}

func (p *Provider) Descriptor() ports.IdentityProviderDescriptor {
	return ports.IdentityProviderDescriptor{
		ID:          p.id,
		Type:        "oidc",
		DisplayName: p.displayName,
		Issuer:      p.issuer,
		TrustPolicy: model.LoginTrustPolicy{
			EmailBinding:        p.emailBinding,
			AllowedEmailDomains: append([]string(nil), p.allowedEmailDomainList...),
		},
	}
}

func (p *Provider) BeginAuth(ctx context.Context, req ports.BeginAuthRequest) (ports.BeginAuthResult, error) {
	options := []oauth2.AuthCodeOption{oauth2.AccessTypeOnline}
	privateState := oidcPrivateState{}
	if req.Nonce != "" {
		options = append(options, oidc.Nonce(req.Nonce))
	}
	if req.LoginHint != "" {
		options = append(options, oauth2.SetAuthURLParam("login_hint", req.LoginHint))
	}
	if p.pkce == "required" {
		verifier, err := randomVerifier()
		if err != nil {
			return ports.BeginAuthResult{}, err
		}
		privateState.CodeVerifier = verifier
		options = append(options,
			oauth2.SetAuthURLParam("code_challenge", codeChallengeS256(verifier)),
			oauth2.SetAuthURLParam("code_challenge_method", "S256"),
		)
	}
	rawPrivateState, err := json.Marshal(privateState)
	if err != nil {
		return ports.BeginAuthResult{}, err
	}
	if privateState.CodeVerifier == "" {
		rawPrivateState = nil
	}
	return ports.BeginAuthResult{RedirectURL: p.oauth2Config.AuthCodeURL(req.State, options...), PrivateState: rawPrivateState}, nil
}

func (p *Provider) CompleteAuth(ctx context.Context, req ports.CompleteAuthRequest) (model.ExternalIdentity, error) {
	options, err := p.tokenExchangeOptions(req.PrivateState)
	if err != nil {
		return model.ExternalIdentity{}, err
	}
	token, err := p.oauth2Config.Exchange(ctx, req.Code, options...)
	if err != nil {
		return model.ExternalIdentity{}, fmt.Errorf("oidc: exchange code: %w", err)
	}
	rawIDToken, ok := token.Extra("id_token").(string)
	if !ok || rawIDToken == "" {
		return model.ExternalIdentity{}, fmt.Errorf("oidc: oauth response missing id_token")
	}
	idToken, err := p.verifier.Verify(ctx, rawIDToken)
	if err != nil {
		return model.ExternalIdentity{}, fmt.Errorf("oidc: verify id token: %w", err)
	}
	if req.Nonce != "" && idToken.Nonce != req.Nonce {
		return model.ExternalIdentity{}, fmt.Errorf("oidc: id token nonce mismatch")
	}
	var claims map[string]any
	if err := idToken.Claims(&claims); err != nil {
		return model.ExternalIdentity{}, fmt.Errorf("oidc: decode id token claims: %w", err)
	}
	subject, err := p.subject(claims, idToken.Subject)
	if err != nil {
		return model.ExternalIdentity{}, err
	}
	identity := model.ExternalIdentity{ProviderID: p.id, Issuer: idToken.Issuer, Subject: subject, SubjectStrategy: p.subjectStrategy}
	identity.Email = stringClaim(claims, p.claim("email", "email"))
	identity.EmailVerified = boolClaim(claims, p.claim("email_verified", "email_verified"))
	identity.DisplayName = stringClaim(claims, p.claim("name", "name"))
	identity.PictureURL = stringClaim(claims, p.claim("picture", "picture"))
	identity.HostedDomain = stringClaim(claims, p.claim("hosted_domain", "hd"))
	if err := p.validateTrust(identity); err != nil {
		return model.ExternalIdentity{}, err
	}
	return identity, nil
}

func (p *Provider) subject(claims map[string]any, oidcSubject string) (string, error) {
	switch p.subjectStrategy {
	case "", "oidc_sub":
		return oidcSubject, nil
	case "entra_oid_tid":
		tid := stringClaim(claims, "tid")
		if p.requiredTenantID != "" && tid != p.requiredTenantID {
			return "", fmt.Errorf("%w: oidc tenant mismatch", model.ErrPermissionDenied)
		}
		oid := stringClaim(claims, "oid")
		if tid == "" || oid == "" {
			return "", fmt.Errorf("oidc: entra_oid_tid requires tid and oid claims")
		}
		return tid + ":" + oid, nil
	default:
		return "", fmt.Errorf("oidc: unsupported subject strategy %q", p.subjectStrategy)
	}
}

func (p *Provider) validateTrust(identity model.ExternalIdentity) error {
	if p.profile == "google" {
		if err := p.validateHostedDomain(identity); err != nil {
			return err
		}
	}
	return nil
}

func (p *Provider) claim(canonical, fallback string) string {
	if name := strings.TrimSpace(p.claimMapping[canonical]); name != "" {
		return name
	}
	return fallback
}

func (p *Provider) validateHostedDomain(id model.ExternalIdentity) error {
	hostedDomain := strings.ToLower(strings.TrimSpace(id.HostedDomain))
	if len(p.allowedHostedDomains) > 0 {
		if _, ok := p.allowedHostedDomains[hostedDomain]; !ok {
			return fmt.Errorf("%w: google hosted domain is not allowed", model.ErrPermissionDenied)
		}
		return nil
	}
	if hostedDomain == "" && p.personalAccounts == "deny" {
		return fmt.Errorf("%w: google personal accounts are not allowed", model.ErrPermissionDenied)
	}
	return nil
}

type oidcPrivateState struct {
	CodeVerifier string `json:"code_verifier,omitempty"`
}

func (p *Provider) tokenExchangeOptions(raw []byte) ([]oauth2.AuthCodeOption, error) {
	if len(raw) == 0 {
		if p.pkce == "required" {
			return nil, fmt.Errorf("oidc: pkce code_verifier missing")
		}
		return nil, nil
	}
	var privateState oidcPrivateState
	if err := json.Unmarshal(raw, &privateState); err != nil {
		return nil, fmt.Errorf("oidc: decode private state: %w", err)
	}
	if privateState.CodeVerifier == "" {
		if p.pkce == "required" {
			return nil, fmt.Errorf("oidc: pkce code_verifier missing")
		}
		return nil, nil
	}
	return []oauth2.AuthCodeOption{oauth2.SetAuthURLParam("code_verifier", privateState.CodeVerifier)}, nil
}

func copyClaimMapping(in map[string]string) map[string]string {
	out := make(map[string]string, len(in))
	for key, value := range in {
		out[strings.TrimSpace(key)] = strings.TrimSpace(value)
	}
	return out
}

func normalizeDomains(values []string) map[string]struct{} {
	return domainsToSet(normalizeDomainList(values))
}

func normalizeDomainList(values []string) []string {
	seen := map[string]struct{}{}
	list := make([]string, 0, len(values))
	for _, value := range values {
		value = strings.ToLower(strings.TrimSpace(value))
		if value == "" {
			continue
		}
		if _, ok := seen[value]; ok {
			continue
		}
		seen[value] = struct{}{}
		list = append(list, value)
	}
	return list
}

func domainsToSet(values []string) map[string]struct{} {
	out := make(map[string]struct{}, len(values))
	for _, value := range values {
		out[value] = struct{}{}
	}
	return out
}

func stringClaim(claims map[string]any, name string) string {
	value, _ := claims[name].(string)
	return value
}

func boolClaim(claims map[string]any, name string) bool {
	switch value := claims[name].(type) {
	case bool:
		return value
	case string:
		return strings.EqualFold(strings.TrimSpace(value), "true")
	default:
		return false
	}
}

func displayNameOrDefault(displayName, providerID string) string {
	if displayName = strings.TrimSpace(displayName); displayName != "" {
		return displayName
	}
	return providerID
}

func defaultString(value, fallback string) string {
	if value = strings.TrimSpace(value); value != "" {
		return value
	}
	return fallback
}

func randomVerifier() (string, error) {
	raw := make([]byte, 32)
	if _, err := rand.Read(raw); err != nil {
		return "", fmt.Errorf("oidc: generate pkce verifier: %w", err)
	}
	return base64.RawURLEncoding.EncodeToString(raw), nil
}

func codeChallengeS256(verifier string) string {
	sum := sha256.Sum256([]byte(verifier))
	return base64.RawURLEncoding.EncodeToString(sum[:])
}

File: crates/lore-auth-adapters/src/oidc.rs
//! OIDC identity provider adapter implementations.

use std::{
    collections::{BTreeMap, BTreeSet},
    sync::LazyLock,
};

use async_trait::async_trait;
use base64::{Engine as _, engine::general_purpose::URL_SAFE_NO_PAD};
use lore_auth_core::{
    CoreError,
    model::{ExternalIdentity, LoginTrustPolicy},
    ports::{
        BeginAuthRequest, BeginAuthResult, CompleteAuthRequest, IdentityProvider,
        IdentityProviderDescriptor,
    },
};
use openidconnect::{
    AuthorizationCode, ClientId, ClientSecret, CsrfToken, IssuerUrl, Nonce, PkceCodeChallenge,
    PkceCodeVerifier, RedirectUrl, Scope, TokenResponse,
    core::{CoreAuthenticationFlow, CoreClient, CoreProviderMetadata},
    reqwest,
};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

use crate::config::IdentityProviderConfig;

static DEFAULT_SCOPES: LazyLock<Vec<String>> = LazyLock::new(|| {
    vec![
        "openid".to_owned(),
        "email".to_owned(),
        "profile".to_owned(),
    ]
});

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Config {
    pub provider_id: String,
    pub profile: String,
    pub display_name: String,
    pub issuer: String,
    pub client_id: String,
    pub client_secret: String,
    pub redirect_url: String,
    pub scopes: Vec<String>,
    pub claim_mapping: BTreeMap<String, String>,
    pub subject_strategy: String,
    pub required_tenant_id: String,
    pub email_binding: String,
    pub pkce: String,
    pub allowed_email_domains: Vec<String>,
    pub allowed_hosted_domains: Vec<String>,
    pub personal_accounts: String,
}

impl Config {
    #[must_use]
    pub fn from_provider_config(
        provider_id: impl Into<String>,
        provider: &IdentityProviderConfig,
        client_secret: impl Into<String>,
    ) -> Self {
        Self {
            provider_id: provider_id.into(),
            profile: provider.profile.clone(),
            display_name: provider.display_name.clone(),
            issuer: provider.issuer.clone(),
            client_id: provider.client_id.clone(),
            client_secret: client_secret.into(),
            redirect_url: provider.redirect_url.clone(),
            scopes: provider.scopes.clone(),
            claim_mapping: provider.claims.clone(),
            subject_strategy: provider.subject.strategy.clone(),
            required_tenant_id: provider.subject.required_tid.clone(),
            email_binding: provider.trust.email_binding.clone(),
            pkce: provider.pkce.clone(),
            allowed_email_domains: provider.trust.allowed_email_domains.clone(),
            allowed_hosted_domains: provider.trust.hosted_domain.allowed.clone(),
            personal_accounts: provider.trust.personal_accounts.clone(),
        }
    }
}

#[derive(Clone, Debug)]
pub struct Provider {
    id: String,
    profile: String,
    display_name: String,
    issuer: String,
    client_id: String,
    client_secret: String,
    redirect_url: String,
    scopes: Vec<String>,
    provider_metadata: CoreProviderMetadata,
    http_client: reqwest::Client,
    claim_mapping: BTreeMap<String, String>,
    subject_strategy: String,
    required_tenant_id: String,
    email_binding: String,
    pkce: String,
    allowed_email_domains: Vec<String>,
    allowed_hosted_domains: BTreeSet<String>,
    personal_accounts: String,
}

impl Provider {
    pub async fn discover(config: Config) -> Result<Self, CoreError> {
        let config = normalized_config(config)?;
        let http_client = reqwest::ClientBuilder::new()
            .redirect(reqwest::redirect::Policy::none())
            .build()
            .map_err(|err| CoreError::InvalidArgument(format!("oidc: build http client: {err}")))?;
        let issuer_url = IssuerUrl::new(config.issuer.clone()).map_err(|err| {
            CoreError::InvalidArgument(format!("oidc: issuer is not a valid URL: {err}"))
        })?;
        let provider_metadata = CoreProviderMetadata::discover_async(issuer_url, &http_client)
            .await
            .map_err(|err| {
                CoreError::InvalidArgument(format!(
                    "oidc: discover provider {:?}: {err}",
                    config.issuer
                ))
            })?;
        Self::from_metadata(config, provider_metadata, http_client)
    }

    pub fn from_metadata(
        config: Config,
        provider_metadata: CoreProviderMetadata,
        http_client: reqwest::Client,
    ) -> Result<Self, CoreError> {
        let config = normalized_config(config)?;
        // Keep the issuer exactly as advertised; Url printing would append a
        // trailing slash to bare-origin issuers and diverge from the Go form.
        let issuer = provider_metadata.issuer().as_str().to_owned();
        Ok(Self {
            id: config.provider_id.clone(),
            profile: config.profile,
            display_name: display_name_or_default(&config.display_name, &config.provider_id),
            issuer,
            client_id: config.client_id,
            client_secret: config.client_secret,
            redirect_url: config.redirect_url,
            scopes: if config.scopes.is_empty() {
                DEFAULT_SCOPES.clone()
            } else {
                config.scopes
            },
            provider_metadata,
            http_client,
            claim_mapping: copy_claim_mapping(config.claim_mapping),
            subject_strategy: default_string(&config.subject_strategy, "oidc_sub"),
            required_tenant_id: config.required_tenant_id.trim().to_owned(),
            email_binding: default_string(&config.email_binding, "disabled"),
            pkce: config.pkce.trim().to_owned(),
            allowed_email_domains: normalize_domain_list(config.allowed_email_domains),
            allowed_hosted_domains: normalize_domain_list(config.allowed_hosted_domains)
                .into_iter()
                .collect(),
            personal_accounts: config.personal_accounts.trim().to_owned(),
        })
    }

    fn redirect_url(&self) -> Result<RedirectUrl, CoreError> {
        RedirectUrl::new(self.redirect_url.clone()).map_err(|err| {
            CoreError::InvalidArgument(format!("oidc: redirect URL is invalid: {err}"))
        })
    }

    fn claim<'a>(&'a self, canonical: &'a str, fallback: &'a str) -> &'a str {
        self.claim_mapping
            .get(canonical)
            .map(String::as_str)
            .filter(|name| !name.trim().is_empty())
            .unwrap_or(fallback)
    }

    fn subject(
        &self,
        claims: &Map<String, Value>,
        oidc_subject: &str,
    ) -> Result<String, CoreError> {
        match self.subject_strategy.as_str() {
            "" | "oidc_sub" => Ok(oidc_subject.to_owned()),
            "entra_oid_tid" => {
                let tid = string_claim(claims, "tid");
                if !self.required_tenant_id.is_empty() && tid != self.required_tenant_id {
                    return Err(CoreError::PermissionDenied);
                }
                let oid = string_claim(claims, "oid");
                if tid.is_empty() || oid.is_empty() {
                    return Err(CoreError::InvalidArgument(
                        "oidc: entra_oid_tid requires tid and oid claims".to_owned(),
                    ));
                }
                Ok(format!("{tid}:{oid}"))
            }
            strategy => Err(CoreError::InvalidArgument(format!(
                "oidc: unsupported subject strategy {strategy:?}"
            ))),
        }
    }

    fn validate_trust(&self, identity: &ExternalIdentity) -> Result<(), CoreError> {
        if self.profile == "google" {
            self.validate_hosted_domain(identity)?;
        }
        Ok(())
    }

    fn validate_hosted_domain(&self, identity: &ExternalIdentity) -> Result<(), CoreError> {
        let hosted_domain = identity.hosted_domain.trim().to_lowercase();
        if !self.allowed_hosted_domains.is_empty() {
            if !self.allowed_hosted_domains.contains(&hosted_domain) {
                return Err(CoreError::PermissionDenied);
            }
            return Ok(());
        }
        if hosted_domain.is_empty() && self.personal_accounts == "deny" {
            return Err(CoreError::PermissionDenied);
        }
        Ok(())
    }

    fn token_exchange_verifier(
        &self,
        raw_private_state: &[u8],
    ) -> Result<Option<PkceCodeVerifier>, CoreError> {
        if raw_private_state.is_empty() {
            if self.pkce == "required" {
                return Err(CoreError::InvalidArgument(
                    "oidc: pkce code_verifier missing".to_owned(),
                ));
            }
            return Ok(None);
        }
        let private_state: PrivateState =
            serde_json::from_slice(raw_private_state).map_err(|err| {
                CoreError::InvalidArgument(format!("oidc: decode private state: {err}"))
            })?;
        let verifier = private_state.code_verifier.unwrap_or_default();
        if verifier.is_empty() {
            if self.pkce == "required" {
                return Err(CoreError::InvalidArgument(
                    "oidc: pkce code_verifier missing".to_owned(),
                ));
            }
            return Ok(None);
        }
        Ok(Some(PkceCodeVerifier::new(verifier)))
    }
}

#[async_trait]
impl IdentityProvider for Provider {
    fn descriptor(&self) -> IdentityProviderDescriptor {
        IdentityProviderDescriptor {
            id: self.id.clone(),
            provider_type: "oidc".to_owned(),
            display_name: self.display_name.clone(),
            issuer: self.issuer.clone(),
            trust_policy: LoginTrustPolicy {
                email_binding: self.email_binding.clone(),
                allowed_email_domains: self.allowed_email_domains.clone(),
            },
        }
    }

    async fn begin_auth(&self, req: BeginAuthRequest) -> Result<BeginAuthResult, CoreError> {
        let client = CoreClient::from_provider_metadata(
            self.provider_metadata.clone(),
            ClientId::new(self.client_id.clone()),
            Some(ClientSecret::new(self.client_secret.clone())),
        )
        .set_redirect_uri(self.redirect_url()?);
        let state = req.state.clone();
        let nonce = req.nonce.clone();
        let mut auth_request = client.authorize_url(
            CoreAuthenticationFlow::AuthorizationCode,
            || CsrfToken::new(state),
            || Nonce::new(nonce),
        );
        for scope in &self.scopes {
            // openidconnect always requests the "openid" scope itself; adding
            // it again would send a duplicated scope value on the wire.
            if scope != "openid" {
                auth_request = auth_request.add_scope(Scope::new(scope.clone()));
            }
        }
        if !req.login_hint.is_empty() {
            auth_request = auth_request.add_extra_param("login_hint", req.login_hint);
        }

        let mut private_state = PrivateState::default();
        if self.pkce == "required" {
            let (challenge, verifier) = PkceCodeChallenge::new_random_sha256();
            private_state.code_verifier = Some(verifier.secret().clone());
            auth_request = auth_request.set_pkce_challenge(challenge);
        }

        let (mut redirect_url, _, _) = auth_request.url();
        if req.nonce.is_empty() {
            remove_query_param(&mut redirect_url, "nonce");
        }
        let raw_private_state = if private_state.code_verifier.is_some() {
            serde_json::to_vec(&private_state).map_err(|err| {
                CoreError::InvalidArgument(format!("oidc: encode private state: {err}"))
            })?
        } else {
            Vec::new()
        };
        Ok(BeginAuthResult {
            redirect_url: redirect_url.to_string(),
            private_state: raw_private_state,
        })
    }

    async fn complete_auth(&self, req: CompleteAuthRequest) -> Result<ExternalIdentity, CoreError> {
        let client = CoreClient::from_provider_metadata(
            self.provider_metadata.clone(),
            ClientId::new(self.client_id.clone()),
            Some(ClientSecret::new(self.client_secret.clone())),
        )
        .set_redirect_uri(self.redirect_url()?);
        let mut token_request = client
            .exchange_code(AuthorizationCode::new(req.code.clone()))
            .map_err(|err| CoreError::InvalidArgument(format!("oidc: exchange code: {err}")))?;
        if let Some(verifier) = self.token_exchange_verifier(&req.private_state)? {
            token_request = token_request.set_pkce_verifier(verifier);
        }
        let token_response = token_request
            .request_async(&self.http_client)
            .await
            .map_err(|_| CoreError::Unauthenticated)?;
        let id_token = token_response
            .id_token()
            .ok_or(CoreError::Unauthenticated)?;
        let raw_id_token = id_token.to_string();
        let verifier = client.id_token_verifier();
        let expected_nonce = req.nonce.clone();
        let verified_claims = id_token
            .claims(&verifier, |actual_nonce: Option<&Nonce>| {
                if expected_nonce.is_empty() {
                    return Ok(());
                }
                match actual_nonce {
                    Some(actual) if actual.secret() == &expected_nonce => Ok(()),
                    _ => Err("oidc: id token nonce mismatch".to_owned()),
                }
            })
            .map_err(|_| CoreError::Unauthenticated)?;
        let raw_claims = jwt_claims(&raw_id_token)?;
        let subject = self.subject(&raw_claims, verified_claims.subject().as_str())?;
        let identity = ExternalIdentity {
            provider_id: self.id.clone(),
            issuer: verified_claims.issuer().as_str().to_owned(),
            subject,
            subject_strategy: self.subject_strategy.clone(),
            email: string_claim(&raw_claims, self.claim("email", "email")),
            email_verified: bool_claim(&raw_claims, self.claim("email_verified", "email_verified")),
            display_name: string_claim(&raw_claims, self.claim("name", "name")),
            picture_url: string_claim(&raw_claims, self.claim("picture", "picture")),
            hosted_domain: string_claim(&raw_claims, self.claim("hosted_domain", "hd")),
            ..ExternalIdentity::default()
        };
        self.validate_trust(&identity)?;
        Ok(identity)
    }
}

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
struct PrivateState {
    #[serde(skip_serializing_if = "Option::is_none")]
    code_verifier: Option<String>,
}

fn normalized_config(mut config: Config) -> Result<Config, CoreError> {
    config.provider_id = config.provider_id.trim().to_owned();
    if config.provider_id.is_empty() {
        return Err(CoreError::InvalidArgument(
            "oidc: provider id is required".to_owned(),
        ));
    }
    config.issuer = config.issuer.trim().to_owned();
    if config.issuer.is_empty() {
        return Err(CoreError::InvalidArgument(
            "oidc: issuer is required".to_owned(),
        ));
    }
    config.client_id = config.client_id.trim().to_owned();
    if config.client_id.is_empty() {
        return Err(CoreError::InvalidArgument(
            "oidc: client id is required".to_owned(),
        ));
    }
    config.profile = config.profile.trim().to_owned();
    config.personal_accounts = config.personal_accounts.trim().to_owned();
    match config.personal_accounts.as_str() {
        "" | "allow" | "deny" => {}
        _ => {
            return Err(CoreError::InvalidArgument(format!(
                "oidc: personal_accounts {:?} is unknown",
                config.personal_accounts
            )));
        }
    }
    if !config.personal_accounts.is_empty() && config.profile != "google" {
        return Err(CoreError::InvalidArgument(
            "oidc: personal_accounts is only valid for google profile".to_owned(),
        ));
    }
    Ok(config)
}

fn copy_claim_mapping(mapping: BTreeMap<String, String>) -> BTreeMap<String, String> {
    mapping
        .into_iter()
        .map(|(key, value)| (key.trim().to_owned(), value.trim().to_owned()))
        .collect()
}

fn normalize_domain_list(values: Vec<String>) -> Vec<String> {
    let mut seen = BTreeSet::new();
    let mut out = Vec::new();
    for value in values {
        let value = value.trim().to_lowercase();
        if value.is_empty() || !seen.insert(value.clone()) {
            continue;
        }
        out.push(value);
    }
    out
}

fn string_claim(claims: &Map<String, Value>, name: &str) -> String {
    claims
        .get(name)
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_owned()
}

fn bool_claim(claims: &Map<String, Value>, name: &str) -> bool {
    match claims.get(name) {
        Some(Value::Bool(value)) => *value,
        Some(Value::String(value)) => value.trim().eq_ignore_ascii_case("true"),
        _ => false,
    }
}

fn display_name_or_default(display_name: &str, provider_id: &str) -> String {
    let display_name = display_name.trim();
    if display_name.is_empty() {
        provider_id.to_owned()
    } else {
        display_name.to_owned()
    }
}

fn default_string(value: &str, fallback: &str) -> String {
    let value = value.trim();
    if value.is_empty() {
        fallback.to_owned()
    } else {
        value.to_owned()
    }
}

fn jwt_claims(raw_id_token: &str) -> Result<Map<String, Value>, CoreError> {
    let payload = raw_id_token
        .split('.')
        .nth(1)
        .ok_or_else(|| CoreError::InvalidArgument("oidc: malformed id token".to_owned()))?;
    let raw = URL_SAFE_NO_PAD
        .decode(payload)
        .map_err(|err| CoreError::InvalidArgument(format!("oidc: decode id token: {err}")))?;
    let value: Value = serde_json::from_slice(&raw).map_err(|err| {
        CoreError::InvalidArgument(format!("oidc: decode id token claims: {err}"))
    })?;
    value.as_object().cloned().ok_or_else(|| {
        CoreError::InvalidArgument("oidc: id token claims must be object".to_owned())
    })
}

fn remove_query_param(url: &mut openidconnect::url::Url, key: &str) {
    let pairs = url
        .query_pairs()
        .filter(|(name, _)| name != key)
        .map(|(name, value)| (name.into_owned(), value.into_owned()))
        .collect::<Vec<_>>();
    url.set_query(None);
    {
        let mut query = url.query_pairs_mut();
        for (name, value) in pairs {
            query.append_pair(&name, &value);
        }
    }
}
